Sming/Libraries/OtaUpgrade/BasicOtaUpgradeStream.h

```diff
@@ -0,0 +1,181 @@
+/****
+ * Sming Framework Project - Open Source framework for high efficiency native ESP8266 development.
+ * Created 2017 by Slavey Karadzhov
+ * http://github.com/SmingHub/Sming
+ * All files of the Sming Core are provided under the LGPL v3 license.
+ *
+ * BasicOtaUpgradeStream.h
+ *
+ ****/
+
+#pragma once
+
+#include <Data/Stream/ReadWriteStream.h>
+#include <rboot-api.h>
+#include "OTA_FileFormat.h"
+#ifdef ENABLE_OTA_SIGNING
+#include "OtaSignatureVerifier.h"
+#else
+#include "OtaChecksumVerifier.h"
+#endif
+
+/**
+ * @brief A write-only stream to parse and apply firmware unencrypted upgrade files generated by otatool.py
+ *
+ * The class fully automates the firmware upgrade process without any manual 
+ * configuration. At construction time, the rBoot configuration is read to determine 
+ * the unused ROM slot which should receive the upgrade.
+ * Just feed the upgrade file content into the `write()` method in arbitrarily 
+ * sized chunks. The relevant portion(s) of the Flash memory (currently only the 
+ * application rom) are updated on the fly as data arrives. When the file is complete
+ * and signature validation (if enabled) was successful, the updated slot is activated
+ * in the rBoot configuration.
+ * Call `hasError()` and/or check the public \c #errorCode member to determine if 
+ * everything went smoothly.
+ * 
+ * For further information on configuration options and the file format, 
+ * refer to the library's documentation.
+ * 
+ * @see `EncryptedOtaUpgradeStream` for encryption support.
+ */
+class BasicOtaUpgradeStream : public ReadWriteStream
+{
+public:
+	BasicOtaUpgradeStream();
+
+	/**
+	 * @brief Error code values
+	 */
+	enum ErrorCode {
+		NoError, ///< No error occured thus far (default value of \c #errorCode if `hasError()` returns false)
+		InvalidFormatError,   ///< Invalid/unsupported upgrade file format
+		UnsupportedDataError, ///< Some content of the upgrade file is not supported by this version of OtaUpgradeStream.
+		DecryptionError,	  ///< Decryption failed. Probably wrong decryption key.
+		NoRomFoundError, ///< The file did not contain a ROM image suitable for the start address of the slot to upgrade.
+		RomTooLargeError,   ///< The contained ROM image does not fit into the application firmware slot.
+		DowngradeError,		///< Attempt to downgrade to older firmware version.
+		VerificationError,  ///< Signature/checksum verification failed - updated ROM not activated
+		FlashWriteError,	///< Error while writing to Flash memory.
+		RomActivationError, ///< Error while activating updated ROM slot.
+		OutOfMemoryError,   ///< Dynamic memory allocation failed
+		InternalError		///< An unexpected error occured.
+	};
```

Comment:
I renamed some error codes, because they sound strange without the "Error" suffix, e. g. `Error::FlashWrite` => `Error::FlashWriteError` (the duplication of "Error" is acceptable, I guess).

Reply:
FlashWriteFailure ?

Reply:
That sounds good too.